src/modules/ellswift/main_impl.h

```diff
@@ -0,0 +1,588 @@
+#endif
+
+    if (!(c & 2)) {
+        /* c is in {0, 1, 4, 5}. In this case we look for an inverse under the x1 (if c=0 or
+         * c=4) formula, or x2 (if c=1 or c=5) formula. */
+
+        /* If -u-x is a valid X coordinate, fail. This would yield an encoding that roundtrips
+         * back under the x3 formula instead (which has priority over x1 and x2, so the decoding
+         * would not match x). */
+        m = x; /* m = x */
+        secp256k1_fe_add(&m, &u); /* m = u+x */
+        secp256k1_fe_negate(&m, &m, 2); /* m = -u-x */
+        /* test if (-u-x) is a valid X coordinate. If so, fail. */
+        if (secp256k1_ge_x_on_curve_var(&m)) return 0;
+
+        /* Let s = -(u^3 + 7)/(u^2 + u*x + x^2) [first part] */
+        secp256k1_fe_sqr(&s, &m); /* s = (u+x)^2 */
+        secp256k1_fe_negate(&s, &s, 1); /* s= -(u+x)^2 */
+        secp256k1_fe_mul(&m, &u, &x); /* m = u*x */
+        secp256k1_fe_add(&s, &m); /* s = -(u^2 + u*x + x^2) */
+
+        /* Note that at this point, s=0 is impossible. If it were the case:
+         *    s = -(u^2 + u*x + x^2) = 0.
+         * => u^2 + u*x + x^2 = 0
+         * => (u + 2*x) * (u^2 + u*x + x^2) = 0
+         * => 2*x^3 + 3*x^2*u + 3*x*u^2 + u^3 = 0
+         * => (x + u)^3 + x^3 = 0
+         * => x^3 = -(x + u)^3
+         * => x^3 + B = (-u - x)^3 + B
+         *
+         * However, We know x^3 + B is square (because x is on the curve) and
+         * that (-u-x)^3 + B is not square (the secp256k1_ge_x_on_curve_var(&m)
+         * test above would have failed). This is a contradiction, and thus the
+         * assumption s=0 is false. */
+#ifdef VERIFY
+        VERIFY_CHECK(!secp256k1_fe_normalizes_to_zero_var(&s));
+#endif
+
+        /* If s is not square, fail. We have not fully computed s yet, but s is square iff
+         * -(u^3+7)*(u^2+u*x+x^2) is square (because a/b is square iff a*b is square and b is
+         * nonzero). */
+        secp256k1_fe_sqr(&g, &u); /* g = u^2 */
+        secp256k1_fe_mul(&g, &g, &u); /* g = u^3 */
+        secp256k1_fe_add_int(&g, SECP256K1_B); /* g = u^3+7 */
+        secp256k1_fe_mul(&m, &s, &g); /* m = -(u^3 + 7)*(u^2 + u*x + x^2) */
+        if (!secp256k1_fe_is_square_var(&m)) return 0;
+
+        /* Let s = -(u^3 + 7)/(u^2 + u*x + x^2) [second part] */
+        secp256k1_fe_inv_var(&s, &s); /* s = -1/(u^2 + u*x + x^2) [cannot be div by 0] */
+        secp256k1_fe_mul(&s, &s, &g); /* s = -(u^3 + 7)/(u^2 + u*x + x^2) */
+
+        /* Let v = x. */
+        v = x;
+    } else {
+        /* c is in {2, 3, 6, 7}. In this case we look for an inverse under the x3 formula. */
+
+        /* Let s = x-u. */
+        secp256k1_fe_negate(&m, &u, 1); /* m = -u */
+        s = m; /* s = -u */
+        secp256k1_fe_add(&s, &x); /* s = x-u */
+
+        /* If s=0, fail. */
+        if (secp256k1_fe_normalizes_to_zero_var(&s)) return 0;
+
+        /* If s is not square, fail. */
+        if (!secp256k1_fe_is_square_var(&s)) return 0;
```

Comment:
Could we drop the `s == 0` check here? I assume it happens with negligible probability on honest inputs, and it's subsumed by the `is_square(q)` check below.

If no:
 * Do we want to add an `EXPECT` to the `s == 0` check? (Not sure if it changes much, but I saw you did the same above in the other function.)
 * Do we want to swap these checks for an earlier return?

Reply:
> > and it's subsumed by the `is_square(q)` check below.
> 
> That's not correct. The tests fail if you comment it out.

Oh.  When I wrote this, I had assumed, for a moment, that we abort if q is a square. But we abort if it's not a square.

Reply:
> I've made these changes (and moved the `s==0` check even further, so that other more-likely-to-fail check go before it).

Perhaps update the pseudocode in the comment accordingly.